### PR TITLE
custom tool handler

### PR DESCRIPTION
**Pull request by alexagr**
Implement "custom tool handler" method - as described in issue #627.

Sample usage:
```
class MyToolMessage(ToolMessage):
    request: str = "my_tool"
    handler: str = "tool_handler"

class MyAgent(ChatAgent):
    def tool_handler(self, message: ToolMessage, tool_name: str, chat_doc: ChatDocument = None) -> str:
        if tool_name == "my_tool":
            # do something

```
I've also modified test_tool_messages and test_tool_messages_async to use the new "custom handler" for some tools.

### CODE REVIEW

File: langroid/agent/base.py [hunk 1]
```diff
@@ -1738,18 +1749,31 @@ async def handle_tool_message_async(
...
             and "chat_doc" in inspect.signature(handler_method).parameters
         )
+        has_tool_name_arg = ("tool_name" in
+                             inspect.signature(handler_method).parameters)
         try:
-            if has_chat_doc_arg:
-                maybe_result = await handler_method(tool, chat_doc=chat_doc)
+            if has_tool_name_arg:
+                if has_chat_doc_arg:
+                    maybe_result = await handler_method(tool, tool_name=tool_name,
```
Comment (pchalasani) [inline langroid/agent/base.py:1768]: `tool_name` is a direct function of `tool`, so it seems redundant to have this arg I think? See my comment in the test

File: tests/main/test_tool_messages_async.py [hunk 2]
```diff
@@ -76,8 +77,16 @@ class MessageHandlingAgent(ChatAgent):
     def file_exists(self, message: FileExistsMessage) -> str:
         return "yes" if message.filename == "requirements.txt" else "no"
 
-    def python_version(self, PythonVersionMessage) -> str:
-        return DEFAULT_PY_VERSION
+    def tool_handler(
+        self,
+        message: ToolMessage,
+        tool_name: str,
```
Comment (pchalasani) [inline tests/main/test_tool_messages_async.py:83]: since we are intending `tool_handler` to be specifically for the `PythonVersionMessage` tool, why introduce `tool_name`? Do you see any scenario where we would want to have this arg?

Comment (alexagr) [reply tests/main/test_tool_messages_async.py]: @pchalasani you are absolutely right! Sorry, my bad. Pushed a fixed version.

File: langroid/agent/base.py (hunk 1 above)
Comment (alexagr) [reply langroid/agent/base.py]: @pchalasani you are absolutely right! Sorry, my bad. Pushed a fixed version.

File: langroid/agent/base.py [hunk 3]
```diff
@@ -278,17 +278,27 @@ def _get_tool_list(
         if not issubclass(message_class, ToolMessage):
             raise ValueError("message_class must be a subclass of ToolMessage")
         tool = message_class.default_value("request")
+
+        """
+        if tool has handler method explicitly defined - use it,
+        otherwise use the tool name as the handler
+        """
+        if hasattr(message_class, "handler"):
```
Comment (pchalasani) [inline langroid/agent/base.py:286]: we should use `_handler` instead of `handler` since we don't want `handler` to be part of the schema passed to the LLM (i.e. we don't want LLM to generate `handler` field). 

Also a note on this should be added to the `chat-agent-tool.md` in the quick-start docs. But let's wait until we know we really need this feature (see comment later)

**Review by pchalasani — CHANGES_REQUESTED** [review]
@alexagr thanks for this PR. I wanted to revisit the motivation for this, which I think I'm not fully understanding. 

Pasting from the related issue, you wrote:

> Langroid currently assumes that tool handlers are class methods with names identical to tool's name (request attribute).
This is somewhat unsafe - 
> - [1] as tool name may be identical to an existing method of Agent class. 
> - [2] Also for use-cases where tools are defined dynamically (e.g. based on some "configuration data" stored in external database) this requires creation of multiple dynamic methods in Agent object - which may be a challenge for garbage collection.

Let's take point [1] - say I have an agent A with method `blah`, then when we create a tool T, we **know** we shouldn't set `request` to `blah`. Also, say there's an existing tool T with method `blah`, then when we create agents A1, A2 to use T, we'll have to take care not to have methods `blah`. I think you're saying these considerations of avoiding method-name-clash are a burden, so we use `handler` method to avoid this? But the exact same considerations apply to `handler`, right?

Point [2] about dynamic tool definitions. Here also whatever concerns there are about name-clash with `request` seem to apply to `handler`.

Overall, I understand you want to separate the notion of "tool name" from "tool handler". I'd like to know a scenario you have in mind where we'd leverage that. 

@nilspalumbo please weigh in when you get a chance.

**Comment by nilspalumbo** [comment]
@alexagr I need a bit more clarification as well.

I assume the intention for [1] is that, if there is a name clash, renaming the handler method and `handler` for all tools is simpler than changing `request` and any handler methods defined on the agent for all tools?

Is the goal for [2] that, rather than dynamically redefining a handler method on the agent, to instead redefine some `handler` method as tools are enabled and disabled, allowing the earlier iterations of `handler` to be garbage collected?

Currently, we do not delete the handler methods created when a tool is enabled, even if handling is later disabled, as the memory usage of handler methods will be negligible in typical use-cases. As the size of the required `handler` method would grow with the function count, the potential memory savings are limited. Perhaps this might become relevant with a combination of batch processing and large-scale automated tool generation; is this your intended use case?

**Comment by nilspalumbo** [comment]
Use of union types in a single tool definition already enables much of this functionality (see the `_get_any_tool_message` method of `ChatAgent`).

**Comment by alexagr** [comment]
@pchalasani @nilspalumbo I was referring to a possible clash with base `Agent` class methods - e.g. `agent_response`. 

But let me describe my use-case - as it'll be easier to understand where I'm coming from. I'm writing a Web portal where end-users define agents (prompts / LLMs / ...), tools (URLs / authentication / schema) and upload documents. I'm using langroid as an "engine" to run the chat sessions following these definitions. With the current tools implementation I can't simply create `MyAgent` class with some generic `tool_handler` method - instead I have to create classes dynamically for every session (because configuration changes all the time). Also I can't just use the tool name "as provided by user" - due to the possible naming clash - instead I'm currently adding "tools_" prefix to it - which is pretty messy, since this modified tool name appears in "agent logs" that I expose to users for troubleshooting purposes. Also I'm not sure how well garbage collector cleans up all this mess - though I may be wrong here...

Bottom line - I think that (optional) separation between tool name (as "exposed" to LLM) and handler name (that is "dictated" by underlying code structure) could be quite useful.

**Comment by pchalasani** [comment]
@alexagr I discussed with @nilspalumbo and we decide to proceed with this, since it's a minor change, albeit a bit custom. It will help if you could please:

- add a short illustrative example for how it helps to use the `_handler` attribute in a `ToolMessage`, and 
- add a brief note at `docs/notes/tool_message_handlers.md` and link the example to this note.

**Comment by alexagr** [comment]
@pchalasani thank you - done

**Comment by pchalasani** [comment]
Thanks @alexagr that is a nice illustrative example! Closing as replaced by #643  (I added some sentences to the docs)
